### PR TITLE
Make Pinot JDK 11 Compilable

### PR DESCRIPTION
**Pull request by elonazoulay**
## Description
As part of effort to upgrade Pinot to use JAVA 11(#6689), this PR will:

- Upgrade pinot to be compatible with jdk11+.
This involves updating dependencies, tests and minor modifications to some classes:
equality methods for Schema and ColumnMetadata.
- JDK 8 build is still available with the option: `-Djdk.version=8`
```
mvn clean install -DskipTests -Pbin-dist -T 4  -Djdk.version=8
```

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ X] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ X] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes

Upgrade to Java 11 and drop support for Java 8

If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release.

If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java [hunk 1]
```diff
@@ -36,18 +36,9 @@
...
-
-    } catch (NoSuchMethodException e) {
-      //this should never happen
-      ExceptionUtils.rethrow(e);
-    }
-    method.setAccessible(true);
     for (URL url : urls) {
       try {
-        method.invoke(classLoader, url);
+        super.add(url);
```
Comment (fx19880617) [inline pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java:41]: what's the behavior for java8 on this?

**Comment by fx19880617** [comment]
delete this gc.log.*?

File: kubernetes/helm/pinot/values.yaml [hunk 2]
```diff
@@ -154,7 +154,7 @@ server:
     storageClass: ""
     #storageClass: "ssd"
 
-  jvmOpts: "-Xms512M -Xmx1G -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -Xloggc:/opt/pinot/gc-pinot-server.log"
+  jvmOpts: "-Xms512M -Xmx1G -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -Xlog:gc*,safepoint:gc.log:time,uptime -Xlog:gc:/opt/pinot/gc-pinot-server.log"
```
Comment (fx19880617) [inline kubernetes/helm/pinot/values.yaml:157]: should also change on broker jvmOpts?

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java (hunk 1 above)
Comment (fx19880617) [reply pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java:41]: I think we need to check on jdk version and choose different APIs here.

**Comment by Jackie-Jiang** [comment]
What are we trying to achieve here? Should we provide multiple docker files for different java versions?

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java (hunk 1 above)
Comment (elonazoulay) [reply pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java:41]: Will do!

Comment (elonazoulay) [reply pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java:41]: Update: In java9+ the application class loader no longer extends URIClassLoader: https://community.oracle.com/tech/developers/discussion/4011800/base-classloader-no-longer-from-urlclassloader

Invoking addURL will work for all jdk versions. In the findClass method, if the class is not in the parent class loader it will be in this one. I added a comment to explain.

Also, this has been working for us in production.

**Comment by elonazoulay** [comment]
> What are we trying to achieve here? Should we provide multiple docker files for different java versions?

It seems that java11+ should be used. Let me know if there is any usecase where java10 should be added back (I removed java 1.8 and 10 from the versions to be tested).

**Comment by xiangfu0** [comment]
> > What are we trying to achieve here? Should we provide multiple docker files for different java versions?
> 
> It seems that java11+ should be used. Let me know if there is any usecase where java10 should be added back (I removed java 1.8 and 10 from the versions to be tested).

I think it's ok to bypass JDK 9 and 10.

**Review by xiangfu0 — APPROVED** [review]
+100 for this change!

**Comment by xiangfu0** [comment]
Can you also try to remove this `--add-exports java.base/jdk.internal.ref=ALL-UNNAMED` flag in file: `pinot-tools/src/main/resources/appAssemblerScriptTemplate` to see if it can pass the ci?

**Comment by elonazoulay** [comment]
> Can you also try to remove this `--add-exports java.base/jdk.internal.ref=ALL-UNNAMED` flag in file: `pinot-tools/src/main/resources/appAssemblerScriptTemplate` to see if it can pass the ci?
> With your change I think we can safely remove this tag.

Trying it now!

**Comment by kishoreg** [comment]
pardon my ignorance - why do we say that this change makes it backward-incompatible?

**Comment by elonazoulay** [comment]
> pardon my ignorance - why do we say that this change makes it backward-incompatible?

By backward incompatible do you mean that an older pinot cluster cannot be upgraded? 
If so I would say this is not backward-incompatible: we were able to upgrade a cluster in place with the same
configuration. 

@xiangfu0 do you have some more context?

**Comment by xiangfu0** [comment]
> > pardon my ignorance - why do we say that this change makes it backward-incompatible?
> 
> By backward incompatible do you mean that an older pinot cluster cannot be upgraded?
> If so I would say this is not backward-incompatible: we were able to upgrade a cluster in place with the same
> configuration.
> 
> @xiangfu0 do you have some more context?

I think the issue is that JDK11 built binary cannot be ran using JDK8, which requires all existing JDK8 users to upgrade their environment.

**Comment by xiangfu0** [comment]
Another thing we can try is to add the `jdk.version` override.
E.g. 
```
mvn clean install -Pbin-dist  -DskipTests -Djdk.version=8
```
Then users may still be able to use JDK 8 to build and run Pinot.

However it may result we need to release two versions of Pinots based on JDK.

**Comment by xiangfu0** [comment]
I tested on JDK 8 build, seems that it can work, we just need to change the plugin loader step to have similar behavior as the JDK9+.

Also not yet test the behavior for Hadoop and Spark. It may require putting all the jars into the classpath.

**Comment by xiangfu0** [comment]
Updated the PR to make it compilable with JAVA 8.
Make tests to use java 11 and quickstart can be built and ran by java 8/10

File: pinot-clients/pinot-java-client/src/test/java/org/apache/pinot/client/ExternalViewReaderTest.java [hunk 3]
```diff
@@ -75,7 +75,7 @@ public void testGetLiveBrokersExceptionState() throws IOException {
     // Setup
     final List<String> expectedResult = Arrays.asList();
     when(mockZkClient.readData(Mockito.anyString(), Mockito.anyBoolean())).thenThrow(
-        IOException.class);
+        RuntimeException.class);
```
Comment (Jackie-Jiang) [inline pinot-clients/pinot-java-client/src/test/java/org/apache/pinot/client/ExternalViewReaderTest.java:78]: Is this related?

File: pinot-common/src/test/java/org/apache/pinot/common/utils/PinotDataTypeTest.java [hunk 4]
```diff
@@ -105,6 +105,7 @@ public void testBytes() {
   @Test
   public void testTimestamp() {
     Timestamp timestamp = Timestamp.valueOf(LocalDateTime.now());
+    timestamp.setNanos(0);
```
Comment (Jackie-Jiang) [inline pinot-common/src/test/java/org/apache/pinot/common/utils/PinotDataTypeTest.java:108]: Is this related?

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java [hunk 5]
```diff
@@ -36,18 +35,16 @@
...
       try {
-        method.invoke(classLoader, url);
+        /**
+         * ClassLoader in java9+ does not extend URLClassLoader.
+         * If the class is not found in the parent classloader,
+         * it will be found in this classloader via findClass().
+         *
+         * @see https://community.oracle.com/tech/developers/discussion/4011800/base-classloader-no-longer-from-urlclassloader
+         */
+        super.addURL(url);
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java:47]: The try-catch can be removed
```suggestion
        addURL(url);
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java [hunk 6]
```diff
@@ -36,18 +35,16 @@
...
-    } catch (NoSuchMethodException e) {
-      //this should never happen
-      ExceptionUtils.rethrow(e);
-    }
-    method.setAccessible(true);
     for (URL url : urls) {
       try {
-        method.invoke(classLoader, url);
+        /**
+         * ClassLoader in java9+ does not extend URLClassLoader.
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java:41]: Don't quite understand the comment here. Do you mean `URLClassLoader` does not extend `ClassLoader` in java 9? But it should still have the method `addURL()` right?

File: kubernetes/helm/pinot/values.yaml [hunk 7]
```diff
@@ -152,7 +152,7 @@ broker:
     # fsGroup: 2000
   securityContext: {}
 
-  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -Xloggc:/opt/pinot/gc-pinot-broker.log"
+  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -Xlog:gc*,safepoint:gc.log:time,uptime -Xlog:gc:/opt/pinot/gc-pinot-broker.log"
```
Comment (Jackie-Jiang) [inline kubernetes/helm/pinot/values.yaml:155]: Do we want to remove the `-XX:MaxGCPauseMillis=200`?

File: kubernetes/helm/pinot/values.yaml [hunk 8]
```diff
@@ -317,7 +317,7 @@ minion:
     readinessEnabled: false
 
   dataDir: /var/pinot/minion/data
-  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -Xloggc:/opt/pinot/gc-pinot-minion.log"
+  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -Xlog:gc*,safepoint:gc.log:time,uptime -Xlog:gc:/opt/pinot/gc-pinot-minion.log"
```
Comment (Jackie-Jiang) [inline kubernetes/helm/pinot/values.yaml:320]: Same here

**Review by Jackie-Jiang — COMMENTED** [review]
Do we want to consider running one test suite with java 8 and another one with java 11 to ensure both work as expected?

File: pinot-common/src/test/java/org/apache/pinot/common/utils/PinotDataTypeTest.java (hunk 4 above)
Comment (xiangfu0) [reply pinot-common/src/test/java/org/apache/pinot/common/utils/PinotDataTypeTest.java:108]: yes, java11 uses nano by default, Pinot converter doesn't keep the nano part.

File: pinot-clients/pinot-java-client/src/test/java/org/apache/pinot/client/ExternalViewReaderTest.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-clients/pinot-java-client/src/test/java/org/apache/pinot/client/ExternalViewReaderTest.java:78]: This is caused by type inference check in java 11 as ZkClient.readData() doesn't throw IOException, so the mock cannot throw it as well.

File: pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java (hunk 6 above)
Comment (xiangfu0) [reply pinot-spi/src/main/java/org/apache/pinot/spi/plugin/PluginClassLoader.java:41]: `URLClassLoader` still extend `ClassLoader` and this `addURL` method is there.
Actually, this part also works for java 8. 
@kishoreg @elonazoulay thoughts?

File: kubernetes/helm/pinot/values.yaml [hunk 9]
```diff
@@ -83,7 +83,7 @@ controller:
     host: pinot-controller
     port: 9000
 
-  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -Xloggc:/opt/pinot/gc-pinot-controller.log"
+  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -Xlog:gc*,safepoint:gc.log:time,uptime -Xlog:gc:/opt/pinot/gc-pinot-controller.log"
```
Comment (Jackie-Jiang) [inline kubernetes/helm/pinot/values.yaml:86]: Remove the extra `-XX:+UseG1GC`

File: kubernetes/helm/pinot/values.yaml [hunk 10]
```diff
@@ -152,7 +152,7 @@ broker:
     # fsGroup: 2000
   securityContext: {}
 
-  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -Xloggc:/opt/pinot/gc-pinot-broker.log"
+  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -Xlog:gc*,safepoint:gc.log:time,uptime -Xlog:gc:/opt/pinot/gc-pinot-broker.log"
```
Comment (Jackie-Jiang) [inline kubernetes/helm/pinot/values.yaml:155]: Same here

File: kubernetes/helm/pinot/values.yaml [hunk 11]
```diff
@@ -317,7 +317,7 @@ minion:
     readinessEnabled: false
 
   dataDir: /var/pinot/minion/data
-  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -Xloggc:/opt/pinot/gc-pinot-minion.log"
+  jvmOpts: "-Xms256M -Xmx1G -XX:+UseG1GC -XX:+UseG1GC -XX:MaxGCPauseMillis=200 -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -Xlog:gc*,safepoint:gc.log:time,uptime -Xlog:gc:/opt/pinot/gc-pinot-minion.log"
```
Comment (Jackie-Jiang) [inline kubernetes/helm/pinot/values.yaml:320]: Same here

File: pinot-common/src/test/java/org/apache/pinot/common/utils/PinotDataTypeTest.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-common/src/test/java/org/apache/pinot/common/utils/PinotDataTypeTest.java:108]: In that case, maybe cleaner if directly initialize with millis
```suggestion
    Timestamp timestamp = new Timestamp(System.currentTimeMillis());
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM. My only concern is how do we ensure the tests can pass in java 8? We still want to keep java 8 supported

**Comment by xiangfu0** [comment]
> LGTM. My only concern is how do we ensure the tests can pass in java 8? We still want to keep java 8 supported

We want to keep java 8 support, so we keep the quickstart test for java 8, but no unit tests or integration tests.

Also, this setup blocks using the code feature of java 11.

**Comment by Jackie-Jiang** [comment]
> > LGTM. My only concern is how do we ensure the tests can pass in java 8? We still want to keep java 8 supported
> 
> We want to keep java 8 support, so we keep the quickstart test for java 8, but no unit tests or integration tests.
> 
> Also, this setup blocks using the code feature of java 11.

If we use the code feature of java 11, we can no longer support compiling in java 8 right?

**Comment by snleee** [comment]
@Jackie-Jiang @xiangfu0 As Jackie pointed out, we cannot compile the code with Java 8 once we start to use java 11 features. Let's announce that we will officially deprecate the java 8 support from `0.8.0 (next)` release. After we cut that release, we can start to use java 11 features.

**Comment by xiangfu0** [comment]
FYI: @yupeng9 @chenboat do you guys have any thoughts?

**Comment by chenboat** [comment]
> FYI: @yupeng9 @chenboat do you guys have any thoughts?

we are in the process of releasing 0.7 now. if java 11 is used from 0.8 onward, we will do some test on our end before our next release.

**Comment by xiangfu0** [comment]
> > FYI: @yupeng9 @chenboat do you guys have any thoughts?
> 
> we are in the process of releasing 0.7 now. if java 11 is used from 0.8 onward, we will do some test on our end before our next release.

We will keep 0.8 still java8 compilable, then drop the support.

**Comment by snleee** [comment]
@chenboat the most recently released version is `0.7.1`. The next release will be `0.8.0`. So, we support java 8 compatible up to `0.8.0` and then drop the support moving forward. How does that sound? Also, is your runtime env using jvm 8 or jvm 11?

**Comment by xiangfu0** [comment]
Discussed with @mcvsubbu @siddharthteotia @jackjlli offline, we will keep the JDK 8 support longer for LinkedIn to test and validate their deployment and data ingestion workflow.

The concrete date of dropping support for JDK8 is still TBD. @snleee
